src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11SecretKeyFactory.java

```diff
@@ -549,34 +587,23 @@ private static P11Key createKey(Token token, byte[] encoded,
         long keyType = ki.keyType;
         try {
             switch ((int) keyType) {
-                case (int) CKK_DES -> {
-                    keyLength =
-                            P11KeyGenerator.checkKeySize(CKM_DES_KEY_GEN, n, token);
-                    fixDESParity(encoded, 0);
-                }
-                case (int) CKK_DES3 -> {
-                    keyLength =
-                            P11KeyGenerator.checkKeySize(CKM_DES3_KEY_GEN, n, token);
-                    fixDESParity(encoded, 0);
-                    fixDESParity(encoded, 8);
-                    if (keyLength == 112) {
-                        keyType = CKK_DES2;
-                    } else {
-                        keyType = CKK_DES3;
-                        fixDESParity(encoded, 16);
+                case (int) CKK_DES, (int) CKK_DES3, (int) CKK_AES, (int) CKK_RC4,
+                        (int) CKK_BLOWFISH, (int) CKK_CHACHA20 -> {
+                    keyLength = P11KeyGenerator.checkKeySize(ki.keyGenMech, n,
+                            token);
+                    if (keyType == CKK_DES || keyType == CKK_DES3) {
+                        fixDESParity(encoded, 0);
+                    }
+                    if (keyType == CKK_DES3) {
+                        fixDESParity(encoded, 8);
+                        if (keyLength == 112) {
+                            keyType = CKK_DES2;
+                        } else {
+                            fixDESParity(encoded, 16);
+                        }
                     }
                 }
```

Comment:
Hmm, how about separating out AES, RC4, Blowfish and ChaCha20 to a separate case? Only DES and DES3 needs parity checking and they are very legacy.

Reply:
We would need to repeat code if we separate (invocation to `P11KeyGenerator::checkKeySize`). Does not look complex enough in my opinion to merit this split.

Reply:
The separation can remove 1 conditional block, so only 1 extra line and the flow looks cleaner in my opinion, e.g.
```suggestion
                case (int) CKK_DES, (int) CKK_DES3 -> {
                    keyLength = P11KeyGenerator.checkKeySize(ki.keyGenMech, n,
                            token);
                    fixDESParity(encoded, 0);
                    if (keyType == CKK_DES3) {
                        fixDESParity(encoded, 8);
                        if (keyLength == 112) {
                            keyType = CKK_DES2;
                        } else {
                            fixDESParity(encoded, 16);
                        }
                    }
                }
                case (int) CKK_AES, (int) CKK_RC4, (int) CKK_BLOWFISH, (int) CKK_CHACHA20 -> {
                    keyLength = P11KeyGenerator.checkKeySize(ki.keyGenMech, n,
                            token);
                }
```

Reply:
If you'd still like lumping them together as you have it now, then at least move the `if (keyType == CKK_DES3)` block (line 624-630) to inside the previous if-block (line 621-623)?